getfileinfo for dirs and appveyor testing

efiop opened the pull request:
Signed-off-by: Ruslan Kuprieiev <kupruser@gmail.com>

sid0 commented:
Thanks so much for the PR! I don't have a Windows build environment at the moment, but I'll try to get one set up to look at this some time later this week.

efiop commented:
@sid0 Hey, if you could enable appveyor(https://www.appveyor.com/ , it is free for opensource and should only take you a few minutes to provide access for it) for your repo, I could set it up myself, so that you have a CI running and testing this stuff. I am using your extremely useful project in ours(dataversioncontrol/dvc) and thus very much interested in ntfsutils being tested and stable.

sid0 commented:
Hey @efiop -- sorry about the delay. I've been intermittent for personal health reasons.

Good point re appveyor! I use it for a couple of other projects. Would you like to set it up for this? I'm happy to give you access to manage the Appveyor installation if you give me your AppVeyor email address. Thanks!

efiop commented:
@sid0 Sorry, didn't mean to hurry you. Hope all is well.

Sure, I am happy to set it up. My email on appveyor is the same as in my github profile(it should be visible at github.com/efiop).

sid0 commented:
@efiop just added you! Let me know how it goes

efiop commented:
@sid0 Thank you! Already working on it, should be ready today.

efiop commented:
@sid0 I've changed the original patch a bit, so it only adds FILE_FLAG_BACKUP_SEMANTICS if path is pointing to the directory. I also added a few simple tests to cover hardlinks and junctions and setup appveyor to run those tests for python 2 and 3. Please take a look when you have time. Thank you :)

sid0 commented on ntfsutils/fs.py:
```diff
@@ -90,7 +91,12 @@ def getfileinfo(path):
     struct of type BY_HANDLE_FILE_INFORMATION.
     """
-    hfile = CreateFile(path, GENERIC_READ, FILE_SHARE_READ, None, OPEN_EXISTING, 0, None)
+    if os.path.isdir(path):
```
So I'm a little bit concerned about [TOCTOU](https://en.wikipedia.org/wiki/Time_of_check_to_time_of_use) races here -- i.e. if something comes in and changes the path in between the `os.path.isdir` call and the `CreateFile` below. Would it be worth doing something like trying to call `CreateFile` without this flag first, then if it fails (and you can see that it's a directory) to try and call it with this flag?

efiop replied:
I totally agree. Updated the code to try once again with FILE_FLAG_BACKUP_SEMANTICS flag. Please take a look.

efiop commented:
On the other hand, maybe it is worth simply creating a getdirinfo() and leaving getfileinfo() intact. What do you think?

sid0 commented:
Yeah, I like the idea to create `getdirinfo()` -- if the user's unsure whether it's a file or a directory they can call both themselves.

efiop commented:
Agreed :+1: . Please see getdirinfo() implemented in the new patch version.

sid0 commented on ntfsutils/fs.py:
```diff
@@ -85,12 +85,8 @@ class BY_HANDLE_FILE_INFORMATION(ctypes.Structure):
-    struct of type BY_HANDLE_FILE_INFORMATION.
-    """
-    hfile = CreateFile(path, GENERIC_READ, FILE_SHARE_READ, None, OPEN_EXISTING, 0, None)
+def _getfileinfo(path, flags=0):
```
small nit: I don't think a default argument is very useful here -- could you have each caller pass in the flags explicitly? That would be a little clearer.

Other than that this looks great. Thank you so much!

efiop replied:
Sure, will update in a second. Thank you!

sid0 commented:
Woo hoo, awesome!

efiop commented:
@sid0 :tada: Thank you!

efiop commented:
@sid0 Ah, one more thing. Could you please release this version, so it is available through pip? No rush though.

sid0 commented:
Just released 0.1.4. Thanks again for your patience.

efiop commented:
@sid0 wow, that was fast! Thank you!
